actors/miner/src/lib.rs

```diff
@@ -2995,6 +2996,301 @@ impl Actor {
+            if !deadline_available_for_compaction(
+                policy,
+                current_deadline.period_start,
+                params.from_deadline,
+                rt.curr_epoch(),
+            ) {
+                // the heavy path: try to do synchronous Window Post verification
+                // current deadline must be in the dispute window to satisfy the condition of synchronous Window POST verification
+                if !deadline_available_for_optimistic_post_dispute(
+                    policy,
+                    current_deadline.period_start,
+                    params.from_deadline,
+                    rt.curr_epoch(),
+                ) {
+                    return Err(actor_error!(
+                        forbidden,
+                        "cannot move from deadline {} when !deadline_available_for_compaction && !deadline_available_for_optimistic_post_dispute",
+                        params.from_deadline
+                    ));
+                }
+
+
+                let dl_from = deadlines
+                    .load_deadline(policy, rt.store(), params.from_deadline)
+                    .map_err(|e| {
+                        e.downcast_default(ExitCode::USR_ILLEGAL_STATE, "failed to load deadline")
+                    })?;
+
+                let proofs_snapshot =
+                    dl_from.optimistic_proofs_snapshot_amt(store).map_err(|e| {
+                        e.downcast_default(
+                            ExitCode::USR_ILLEGAL_STATE,
+                            "failed to load proofs snapshot",
+                        )
+                    })?;
+
+                let partitions_snapshot =
+                    dl_from.partitions_snapshot_amt(store).map_err(|e| {
+                        e.downcast_default(
+                            ExitCode::USR_ILLEGAL_STATE,
+                            "failed to load partitions snapshot",
+                        )
+                    })?;
+
+                // Load sectors for the dispute.
+                let sectors =
+                    Sectors::load(rt.store(), &dl_from.sectors_snapshot).map_err(|e| {
+                        e.downcast_default(
+                            ExitCode::USR_ILLEGAL_STATE,
+                            "failed to load sectors array",
+                        )
+                    })?;
+
+                proofs_snapshot
+                    .for_each(|_, window_proof| {
+                        let mut all_sectors =
+                            Vec::<BitField>::with_capacity(window_proof.partitions.len() as usize);
+                        let mut all_ignored =
+                            Vec::<BitField>::with_capacity(window_proof.partitions.len() as usize);
+
+                        for partition_idx in window_proof.partitions.iter() {
+                            let partition = partitions_snapshot
+                                .get(partition_idx)
```

Comment:
@anorth Our concern is that : after moving partition, the partition id will change, and the dispute may fail for partition id mismatch. That's why we do a through PoST verify to all partitions.

Reply:
That's a good point, removing one partition will re-index all the others. This would break the association with the stored proofs and complicate submitting disputes. These issues are probably fixable, but very complicated.

However, I don't think we can solve this by verifying every proof – it just won't fit in the gas limit at some point, and it's also an overly onerous requirement for the SP to have no faults anywhere in the deadline.

We need to fix it some other way. One idea might be to leave behind an empty partition, so that none are re-indexed. It can be removed during compaction. I think process_deadline_end will already handle this correctly (skipping the partition).

Reply:
Another thing worth mentioning is: according to the comment [here](https://github.com/filecoin-project/builtin-actors/blob/8e82ef71a1858d0530e74e778056dc1b67981587/actors/miner/src/deadline_state.rs#L168-L171), the proof is for multiple partitions, not a single partition, so even some partitions are not being moved, they have to be verified together.

Reply:
Now the implementation only verifies partitions that contains at least one partition being moved, in a single `verify_windowed_post`.

Reply:
Regarding the proof being for multiple partitions: this is confusing and poorly documented, but the single proof byte-array for multiple partitions is in fact a concatenation of multiple proofs, one per partition. You can see some evidence of that [here](https://github.com/filecoin-project/builtin-actors/blob/8e82ef71a1858d0530e74e778056dc1b67981587/actors/miner/src/lib.rs#L583).

This is reflecting the confusing proofs API. But it means we can just slice out the proof bytes (192 each) for exactly the partition that needs to be proven, and we can concatenate multiple partitions' proofs from different submissions into a single array to submit as a single proof to the syscall. I think we should do this to avoid any penalty for batching WindowPoSt.

Note that in your temporary commit, the technique of making an array of proofs won't work. You must concatenate the proof bytes (or call verification multiple times). Array size must be 1.

Unlike disputing, we should _not_ delete the proof after verification. The same proof data may need to be loaded for moving a different partition that was part of the same batch.